### DISCUSSION TITLE
Api: Content version does not return relational data

### DISCUSSION THREAD

**Discussion by joselcvarela** [directus/directus#24695] [Feature Requests]
### Describe the Bug

When trying to retrieve relational data from a version, it only returns the delta from what was saved.
As a consumer, I would expect that I could ask nested data from relationships like other API endpoints

### To Reproduce

First, run Directus with the following Postgres database: [missing-relational-versioning.sql.zip](https://github.com/user-attachments/files/15887807/missing-relational-versioning.sql.zip) You can  use `admin@example.org / password` to login so you are using an authenticated cookie.
Then, access the following address with the session cookie already set:
```
http://localhost:8055/items/example/1?version=next&fields=*,examples.*.*,parent.*,children.*
```

You will see that there are items in the relations list that do not follow the sam structure as the other ones. Those are changes made in the specified version: `next`

### Directus Version

10.12.1 (main)

### Hosting Strategy

Self-Hosted (Custom)

### Database

Postgres image from docker-compose

**Comment by br41nslug** [comment]
What would be your desired behavior here? How it currently works: It fetches the data from the database based on your query and then attempts to it's best ability to merge the saved delta json over that (we currently can't query into the stored delta json)

**Comment by joselcvarela** [comment]
I would expect that the relations saved in a version could retrieve their data as any other item instead of the saved delta.
The delta changes seem that can be already queried using `versionRaw` query parameter.

**Comment by joselcvarela** [comment]
In the example given, here's an excerpt of the response:
```
    "examples": [
      {
        "id": 1,
        "example_id": {
          "id": 1,
          "status": "draft",
          "sort": null,
          "user_created": "939299e2-d6c7-407f-856e-6210b431d328",
          "date_created": "2024-06-18T13:56:40.403Z",
          "user_updated": "939299e2-d6c7-407f-856e-6210b431d328",
          "date_updated": "2024-06-18T13:59:32.360Z",
          "title": "Example 1",
          "parent": 2,
          "examples": [1],
          "children": [2]
        },
        "related_example_id": {
          "id": 2,
          "status": "draft",
          "sort": null,
          "user_created": "939299e2-d6c7-407f-856e-6210b431d328",
          "date_created": "2024-06-18T13:56:44.687Z",
          "user_updated": "939299e2-d6c7-407f-856e-6210b431d328",
          "date_updated": "2024-06-18T13:59:23.273Z",
          "title": "Example 2",
          "parent": 1,
          "examples": [],
          "children": [1]
        }
      },
      {
        "example_id": "1",
        "related_example_id": {
          "id": 3
        }
      }
    ],
```

So, I would expect the last item to have the same fields as other ones, but since it's a change made on the version, it's only returning the delta values.

**Comment by br41nslug** [comment]
Ah yeah looks like that last relation record was added in the delta, and we just merge atm we dont walk the data structure recursively fetching potential extra data / items from the database.

**Comment by burka** [comment]
We will walk into that problem sooner or later, too. It would be great if this could be addressed as we would expect new referenced collections to be provided if we fetch a versions state. Anything else seems like quite inconsistent behavior.

**Comment by rijkvanzanten** [comment]
The original intent of the system / endpoint was to retrieve the soft-saved changes that were made to the item so you can continue on those changes. That was then enhanced by allowing you to query the original item and the changes simultaneously by fetching the original item and then applying the changes. However, that now does mean that supporting full nested fetching on top of the items in the nested static changes becomes tricky, as those are not "real" relational columns in the database (the changes object is just a json blob that are deep merged in). Maybe the true solution here is to explicitly not try to merge item + changes to make it very explicit what's going on. Food for thought

**Comment by jaredtbates** [comment]
I understand the technical limitation here - I'd welcome any suggestions for my team to try as an alternative.

I may be explaining this wrong but from what I understand, @Nabelops is working on a workaround which involves a recursive function on our frontend to get the original item from the Directus item metadata, get the changes, and apply them many layers deep, however it's really cumbersome, fragile, and not something I'd expect us to handle in our codebase really.

The use case for us is that we have a number of "building block" items that are joined to top level "page" items in a M2A relationship to effectively create a flexible WYSIWYG editor with our standard components. However, we need to be able to preview changes to items (and the recursive "building blocks") on our dev site prior to rolling them out to the prod site.

Do you have any suggestions regarding how we could accomplish the draft M2A functionality we're looking for?

**Comment by burka** [comment]
We actively use Content-Versions for our approval workflow, including nested collections. 

It is bad enough, that new nested collections are missed, but not showing any nested changes would ruin our setup. 

We have deeply nested collections and would require to do quite a lot (15+) of nested fetches. It would create quite a lot of remote calls and seems too fragile and slow.

The api to fetch a item with `/items/articles/15?fields=*.*&version=content-version` suggests that it will yield a result in the same form as without the version. Anything else would be against the expectations while reading the api documentation. 

I would expect having that raw form available via the /versions endpoint.

Please, don't remove nested collection fetching.

**Comment by seltzerfish** [comment]
I'm also encountering this issue. Like burka, I also use content versions for approval for website changes for my clients in conjunction with Live Preview as outlined [here](https://docs.directus.io/blog/preview-and-content-versioning-with-nextjs.html). When my clients make changes to nested data in a draft version, I'd like them to be able to see what their site will look like if the draft version gets promoted to the main version, instead of a weird looking delta.

**Comment by Nabelops** [comment]
I want to add some clarity around our use case.  We Directus as a page builder for our site.  Our Brand team utilizes the Directus admin site to edit "Pages"

The pages consist of the page url, SEO information and M2A field of our "building blocks".  It is almost identical to the use case described at https://docs.directus.io/guides/headless-cms/reusable-components.html 

We are coming from another CMS that has a concept of "Published" and "Draft" versions of a content item.  They are used to the workflow of making a new draft of an already published page and being able to review it in our development site while still having the current published version available on the production site.  Once a draft is published, it takes the place of the published item.

In selecting CMS, we had looked for this feature as a requirement since this is our stakeholder's current experience.  We selected Directus knowing that content versioning was already in a beta state and should be ready for us when we launched.  

With the current implementation of content versioning in Directus, we are unable to provide a "preview" of the new version of the content item on our development site since there is no way to get the content model in it's entirety as if the content version was merged.  Due to the M2A relationship of the "building blocks" or page components, when we call `readItem()` with the version key, we don't get the expected response back.

Deleted items are included in the response

![Screenshot 2024-07-29 at 4 17 18 PM](https://github.com/user-attachments/assets/4600cfa7-b41d-4df1-81d1-837874aa79f1)

![Screenshot 2024-07-29 at 4 17 50 PM](https://github.com/user-attachments/assets/e4075073-a5d7-4bfe-9f9b-ad7c0f033f83)

Nested relations are returned in the delta format

![Screenshot 2024-07-29 at 4 18 48 PM](https://github.com/user-attachments/assets/a50e3f79-4de3-4a10-a888-267a4f9f2aac)

Default values of new items are not present

![Screenshot 2024-07-29 at 4 22 25 PM](https://github.com/user-attachments/assets/b17754be-77c5-413f-92c0-207895281947)

![Screenshot 2024-07-29 at 4 23 00 PM](https://github.com/user-attachments/assets/e1d29dc4-89d3-4dbc-84b1-15f6fdc51b31)

I recognize that a lot of these issue arise from the fact that we are using deeper joins and relations.  For SEO purposes, we want all page data at the time of the page load so that our ISR server can have all that information available to crawlers.  This is why we went with the deeper join route.  E.g., `*.*.*.*`.  Also, we hide the building block collection from our Brand team so all content versioning is on the page level.  Without that, it would be hard for them to keep track which building block is on which page.  Not even sure the is a UI representation in the Directus admin for the inverse of M2A.  Aka, to see from a building block which pages it is A2M to.

To reiterate, the goal is to have a complete "page" as if the content version was merged.  Thanks.

**Comment by adiazc9** [comment]
Is there any update in this subject?

**Comment by rijkvanzanten** [comment]
@adiazc9 All updates are in the issue thread above 🙂 Directus' versioning system currently works by saving the delta you created, and then applying those changes when you're querying the version data. That does mean that any nested relational data in those changes aren't actual data records like usual, but are a JSON blob. That in turn makes it trickier to just apply the regular querying engine as it can't fetch with joins etc at the moment

**Comment by burka** [comment]
@rijkvanzanten  I see that it is complicated, but that is what we would love to see, too :-)

Currently it returns changes in nested collections as expected, but not a change to another id, where it would just render the id in the output. Quite unexpected.
Is there any chance to put it on the roadmap? If not we would need to build it our self in a extension, which would make it quite complicated.

**Comment by rijkvanzanten** [comment]
Yeah I'd like to have it too! 🙂 

> Is there any chance to put it on the roadmap?

Yesn't 😅 There's a lot of moving pieces that have to happen first to unblock this; most importantly native support for JSON blobs in querying. Once we can use nested JSON values in a query we can unlock nested joins on the delta stored in the JSON as well. Plan B if that doesn't work as intended is to first query all the data, apply the changes, and then somehow query the nested data again but using the delta as the source of the IDs. Our current querying pipeline heavily relies on nested joins, so that would have to be rearchitected to fit 🤔 

Annnnyyyyways long story short; yes it's high on my todo list, but I'm afraid it'll be a while to be able to properly natively support this in the core platform.

**Comment by burka** [comment]
Thanks for the explanation :-) It would be great to hear any updates on the way :-)

**Comment by 7oque** [comment]
I am currently setting up to use Directus as the CMS I recommend to clients, and bumped into this today after trying to setup live preview with versions as it's a common feature and is available on the other CMS' I'm trying to move people away from so it would be _amazing_ to have it, especially as the live preview panel looks so good.

FWIW I was led to believe this was possible from the docs here: https://github.com/directus/docs/blob/main/content/guides/2.content/5.live-preview.md?plain=1#L26 and it wasn't until I interacted with the api that I realised it returned the delta instead without relations.

**Comment by danilo0311** [comment]
Based on my understanding of the issue, it seems that what you want to achieve is to display the version content just like you do normally, which includes nested M2A collections and M2O, am I right?

I think there's a partial solution to this, but i'm afraid it is a little bit more complex for the M2O relationships.

To normalize the object you attemp to use you could simply do it manually through a custom function like:

```js
export const normalizeVersion = ({ data }: { data: any; }) => {

	/***
	 * * This code needs to be improved and standarized as much as possible
	 * * so it can be used in different scenarios
	 */

	if(!Array.isArray(data) && data) {

		const findTarget = (object: any) => {

			const defaultTargets = Object.keys(object);
			const targets = defaultTargets.filter((action) => action !== 'delete');

			const target = targets.find((target) => target.length !== 0);
			return target;

		};

		const target = findTarget(data);

		switch(target) {

			case 'create': {

				return data[ target ];

			}

			default: {

				return data;

			}

		}

	};

	if(data && data.length !== 0) {

		return data;

	}

	return null;

};
```

Then use the object just like you normally do:

```js
subsections = normalizeVersion({ data: subsections });
translations = normalizeVersion({ data: translations });

const { title } = translation({ translations, lang });
const hasChildren = Array.isArray(subsections) && subsections.length !== 0;
```

I haven't find any solution for the default values not being returned but what I am doing at the moment is just seting them as static values if the variable containing them is undefined, which is only if the user haven't changed the default value explicitely.

For the M2O thing, it does get returned if it existed before creating the new version.
Another thing you'll have to fix is the sort config, in versions you'll have to do it manually as well:

`content.sort((a: any, b: any) => a.sort - b.sort);`

I hope this is useful or at least leds to a better solution than mine.

**Comment by on3dd** [comment]
Any updates on this? It makes versioning in Directus practically useless, at least for us 😕

**Reply by rijkvanzanten** [reply]
Yeah! This is top of our todo list 🙂

**Comment by birdream** [comment]
Hey, any updates on this now? we facing the same issue.

**Comment by pdramos1** [comment]
@rijkvanzanten Just checking in on where on the development list this is.

**Comment by burka** [comment]
We're eagerly waiting for this feature – like kids on Christmas Eve!
How is it going @rijkvanzanten ?

**Reply by rijkvanzanten** [reply]
This is top of the list and currently in development 🥳 https://github.com/directus/directus/pull/25437 is a draft PR of this in flight 💪🏻

**Reply by burka** [reply]
Yeah! Awesome! Love seeing that!

**Reply by mrcjdkhn** [reply]
Do you have already a target release for this feature?

**Reply by rijkvanzanten** [reply]
As soon as possible 😇  @Nitwel found a pretty novel approach to solving this problem relying on database transactions to do the querying realistically, but that does mean we need to take some time to thoroughly test the performance of this to ensure we don't accidentally break high-volume production setups when using this feature. Implementation is pretty far along, so assuming we don't run into a major _this approach can't work because of X_ I'm hopeful this can land in the September 3rd release

**Comment by joselcvarela** [comment]
Seems like this is now included on version [v11.11.0](https://github.com/directus/directus/releases/tag/v11.11.0) with some more improvements on version [v11.12.0](https://github.com/directus/directus/releases/tag/v11.12.0).

Since this is now implemented, I am closing this one out.
Thank you team!
